ts/Stock/Indicators/DMI/DMIIndicator.ts

```diff
@@ -0,0 +1,398 @@
+            period: 14
+        },
+        marker: {
+            enabled: false
+        },
+        tooltip: {
+            pointFormat: '<span style="color:{point.color}">\u25CF</span><b> {series.name}</b><br/>' +
+                '<span style="color:{point.color}">DX</span>: {point.y}<br/>' +
+                '<span style="color:{point.series.options.plusDILine.styles.lineColor}">+DI</span>' +
+                    ': {point.plusDI}<br/>' +
+                '<span style="color:{point.series.options.minusDILine.styles.lineColor}">-DI</span>' +
+                    ': {point.minusDI}<br/>'
+        },
+        /**
+         * +DI line options.
+         */
+        plusDILine: {
+            /**
+             * Styles for the +DI line.
+             */
+            styles: {
+                /**
+                 * Pixel width of the line.
+                 */
+                lineWidth: 1,
+                /**
+                 * Color of the line.
+                 *
+                 * @type {Highcharts.ColorString}
+                 */
+                lineColor: '#00ff00'
+            }
+        },
+        /**
+         * -DI line options.
+         */
+        minusDILine: {
+            /**
+             * Styles for the -DI line.
+             */
+            styles: {
+                /**
+                 * Pixel width of the line.
+                 */
+                lineWidth: 1,
+                /**
+                 * Color of the line.
+                 *
+                 * @type {Highcharts.ColorString}
+                 */
+                lineColor: '#ff0000'
+            }
+        },
+        dataGrouping: {
+            approximation: 'averages'
+        }
+    } as DMIOptions);
+
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    public calculateDM(
+        yVal: Array<Array<number>>,
+        i: number,
+        isPositiveDM?: boolean
+    ): number {
+        const currentHigh = yVal[i][1],
+            currentLow = yVal[i][2],
+            previousHigh = yVal[i - 1][1],
+            previousLow = yVal[i - 1][2];
+
+        // If any value is null, return 0.
+        if (!currentHigh || !currentLow || !previousHigh || !previousLow) {
+            return 0;
+        }
+
+        let DM: number = 0;
+
+        if (currentHigh - previousHigh > previousLow - currentLow) {
+            // for +DM
+            DM = isPositiveDM ? Math.max(currentHigh - previousHigh, 0) : 0;
+        } else {
+            // for -DM
+            DM = !isPositiveDM ? Math.max(previousLow - currentLow, 0) : 0;
+        }
+
+        return correctFloat(DM);
+    }
+
+    public calculateDI(
+        smoothedDM: number,
+        tr: number
+    ): number {
+        return smoothedDM / (tr || 1) * 100;
+    }
+
+    public calculateDX(
+        plusDI: number,
+        minusDI: number
+    ): number {
+        return correctFloat(
+            Math.abs(plusDI - minusDI) / (Math.abs(plusDI + minusDI) || 1) * 100
```

Comment:
Dividing - I want to make it safe from 0, but it might change the indicators values. Well, rather change values than got an error. Is it OK? If not, any ideas?

Reply:
But do we have to really be concerned about that? 
I don't see a chance that the sum of two positive values give 0. 
Both of them must be 0 which is impossible in this case... 
Correct me if I'm wrong?

Reply:
Probably you're right, I was thinking the same. Yea, I'll remove it and, if any errors occur in the future, we can easily fix it.